src/databases/NASTRAN/avtNASTRANFileFormat.C

```diff
@@ -187,90 +187,75 @@ avtNASTRANFileFormat::ActivateTimestep()
 //    Ok, I 'fixed' this funky logic again. The above 'fix' caused the alg.
 //    to basically completely fail.
 //
+//    Mark C. Miller, Sun Jul  4 16:19:34 PDT 2021
+//    Simplified logic for NASTRAN exponential format
 // ****************************************************************************
-
-static float Getf(const char *s)
+static double GetVal(const char *s)
 {
-    char *ends;
-    double val = 0.0;
+    double dval;
+    char *endptr = 0;
+    size_t i;
+
+    // handle a string with no numbers
+    i = strcspn(s, "0123456789");
+    if (s[i] == '\0') return 0;
 
-    // Check for one of these funky 'NASTRAN exponential format' strings.
+    // Check for the funky 'NASTRAN exponential format'.
     // This is where a value like '1.2345e-5' is actually represented in the
     // file as '1.2345-5' with the 'e' character missing. It is awkward but
     // apparently a NASTRAN standard. I guess the rationale is that given
     // an 8 character field width limit, removing the 'e' character gives them
     // one additional digit of precision. This logic is basically looking for
     // the condition of encountering a sign character, '-' or '+', AFTER having
     // seen characters that could represent part of a number. In such a case,
-    // it MUST be the sign of the exponent.
-    const char *p = s;
-    char tmps[32];
-    char *q = tmps;
-    bool haveSeenNumChars = false;
-    while (!haveSeenNumChars || (*p != '-' && *p != '+' && *p != '\0'))
-    {
-        if (('0' <= *p && *p <= '9') || *p == '.' || *p == '+' || *p == '-')
-            haveSeenNumChars = true;
-        *q++ = *p++;
-    }
-    if (haveSeenNumChars && (*p == '-' || *p == '+'))
-    {
-        *q++ = 'e';
-        while (*p != '\0')
-            *q++ = *p++;
-        *q++ = '\0';
-        errno = 0;
-        val = strtod(tmps, &ends);
-    }
-    else
-    {
-        errno = 0;
-        val = strtod(s, &ends);
-    }
+    // it MUST be the sign of an exponent.
+    std::string t(s);
+    for (i = t.size(); t[i]!='-' && t[i]!='+' && t[i]!='\0'; i--);
```

Comment:
Ok, now that I am looking at it, I don't like this logic...it is marching backwards through the string and should end on condition `i==0` and not `t[i]!='\0'`.

Reply:
I fixed this by marching forwards through string but starting at location index 1 (in case there is a leading `+` or `-` sign) because we are looking for a `+` or `-` sign in the *middle* of the string with no `e` or `E` ahead of it.